File: SMILE/tutorials/python/tutorial6.py
import pysmile

# Tutorial6 creates a dynamic Bayesian network (DBN),
# performs the inference, then saves the model to disk.

class Tutorial6:
    def __init__(self):
        print("Starting Tutorial6...")
        net = pysmile.Network()
        loc = self.create_cpt_node(net,
              "Location", "Location",
              ["Pittsburgh", "Sahara"],
              160, 360)
        rain = self.create_cpt_node(net, 
               "Rain", "Rain", 
               ["true", "false"],
               380, 240)
        umb = self.create_cpt_node(net,
              "Umbrella", "Umbrella",
              ["true", "false"],
              300, 100)
        
        net.set_node_temporal_type(rain, pysmile.NodeTemporalType.PLATE)
        net.set_node_temporal_type(umb, pysmile.NodeTemporalType.PLATE)

        net.add_arc(loc, rain)
        net.add_temporal_arc(rain, rain, 1)
        net.add_arc(rain, umb)

        rain_def = [
            0.7,  # P(Rain=true |Location=Pittsburgh)
            0.3,  # P(Rain=false|Location=Pittsburgh)
            0.01, # P(Rain=true |Location=Sahara)
            0.99  # P(Rain=false|Location=Sahara)
        ]

        net.set_node_definition(rain, rain_def)

        rain_def_temporal = [
            0.7,   # P(Rain=true |Location=Pittsburgh,Rain[t-1]=true)
            0.3,   # P(Rain=false|Location=Pittsburgh,Rain[t-1]=true)
            0.3,   # P(Rain=true |Location=Pittsburgh,Rain[t-1]=false)
            0.7,   # P(Rain=false|Location=Pittsburgh,Rain[t-1]=false)
            0.001, # P(Rain=true |Location=Sahara,Rain[t-1]=true)
            0.999, # P(Rain=false|Location=Sahara,Rain[t-1]=true)
            0.01,  # P(Rain=true |Location=Sahara,Rain[t-1]=false)
            0.99   # P(Rain=false|Location=Sahara,Rain[t-1]=false)
        ]
        net.set_node_temporal_definition(rain, 1, rain_def_temporal)

        umb_def = [
            0.9, # P(Umbrella=true |Rain=true)
            0.1, # P(Umbrella=false|Rain=true)
            0.2, # P(Umbrella=true |Rain=false)
            0.8  # P(Umbrella=false|Rain=false)
        ]
        net.set_node_definition(umb, umb_def)

        net.set_slice_count(5)

        print("Performing update without evidence.")
        self.update_and_show_temporal_results(net)

        print("Setting Umbrella[t=1], to true and Umbrella[t=3] to false.")
        net.set_temporal_evidence(umb, 1, 0)
        net.set_temporal_evidence(umb, 3, 1)
        self.update_and_show_temporal_results(net)

        net.write_file("tutorial6.xdsl")
        print("Tutorial6 complete: Network written to tutorial6.xsdl")

    def update_and_show_temporal_results(self, net):
        net.update_beliefs()
        slice_count = net.get_slice_count()
        for h in net.get_all_nodes():
            if net.get_node_temporal_type(h) == pysmile.NodeTemporalType.PLATE:
                outcome_count = net.get_outcome_count(h)
                print("Temporal beliefs for " + net.get_node_id(h) + ":")
                v = net.get_node_value(h)
                for slice_idx in range(0, slice_count):
                    s = "\tt=" + str(slice_idx) + ":"
                    for i in range(0, outcome_count):
                        s = s + " " + str(v[slice_idx * outcome_count + i])
                    print(s)
        print("")

    def create_cpt_node(self, net, id, name, outcomes, x_pos, y_pos):
        handle = net.add_node(pysmile.NodeType.CPT, id)
        
        net.set_node_name(handle, name)
        net.set_node_position(handle, x_pos, y_pos, 85, 55)
        
        initial_outcome_count = net.get_outcome_count(handle)
        
        for i in range(0, initial_outcome_count):
            net.set_outcome_id(handle, i, outcomes[i])
        
        for i in range(initial_outcome_count, len(outcomes)):
            net.add_outcome(handle, outcomes[i])
            
        return handle

File: SMILE/tutorials/csharp/Tutorial6.cs
using System;
using Smile;

// Tutorial6 creates a dynamic Bayesian network (DBN),
// performs the inference, then saves the model to disk.

namespace SmileNetTutorial
{
    class Tutorial6
    {
        public static void Run()
        {
            Console.WriteLine("Starting Tutorial6...");
            Network net = new Network();

            int loc = CreateCptNode(
                net, "Location", "Location",
                new String[] { "Pittsburgh", "Sahara" },
                160, 360);

            int rain = CreateCptNode(
                net, "Rain", "Rain",
                new String[] { "true", "false" },
                380, 240);

            int umb = CreateCptNode(
                net, "Umbrella", "Umbrella",
                new String[] { "true", "false" },
                300, 100);

            net.SetNodeTemporalType(rain, Network.NodeTemporalType.Plate);
            net.SetNodeTemporalType(umb, Network.NodeTemporalType.Plate);

            net.AddArc(loc, rain);
            net.AddTemporalArc(rain, rain, 1);
            net.AddArc(rain, umb);

            double[] rainDef = new double[] 
            {
                0.7,  // P(Rain=true |Location=Pittsburgh)
                0.3,  // P(Rain=false|Location=Pittsburgh)
                0.01, // P(Rain=true |Location=Sahara)
                0.99  // P(Rain=false|Location=Sahara)
            };
            net.SetNodeDefinition(rain, rainDef);

            double[] rainDefTemporal = new double[] 
            {
                0.7,   // P(Rain=true |Location=Pittsburgh,Rain[t-1]=true)
                0.3,   // P(Rain=false|Location=Pittsburgh,Rain[t-1]=true)
                0.3,   // P(Rain=true |Location=Pittsburgh,Rain[t-1]=false)
                0.7,   // P(Rain=false|Location=Pittsburgh,Rain[t-1]=false)
                0.001, // P(Rain=true |Location=Sahara,Rain[t-1]=true)
                0.999, // P(Rain=false|Location=Sahara,Rain[t-1]=true)
                0.01,  // P(Rain=true |Location=Sahara,Rain[t-1]=false)
                0.99  // P(Rain=false|Location=Sahara,Rain[t-1]=false)
		    };
            net.SetNodeTemporalDefinition(rain, 1, rainDefTemporal);

            double[] umbDef = new double[] 
            {
                0.9, // P(Umbrella=true |Rain=true)
                0.1, // P(Umbrella=false|Rain=true)
                0.2, // P(Umbrella=true |Rain=false)
                0.8  // P(Umbrella=false|Rain=false)		
            };
            net.SetNodeDefinition(umb, umbDef);

            net.SetSliceCount(5);

            Console.WriteLine("Performing update without evidence.");
            UpdateAndShowTemporalResults(net);

            Console.WriteLine(
                "Setting Umbrella[t=1] to true and Umbrella[t=3] to false.");
            net.SetTemporalEvidence(umb, 1, 0);
            net.SetTemporalEvidence(umb, 3, 1);
            UpdateAndShowTemporalResults(net);

            net.WriteFile("tutorial6.xdsl");
            Console.WriteLine(
                "Tutorial6 complete: Network written to tutorial6.xdsl");
        }


        private static void UpdateAndShowTemporalResults(Network net)
        {
            net.UpdateBeliefs();
            int sliceCount = net.GetSliceCount();
            for (int h = net.GetFirstNode(); h >= 0; h = net.GetNextNode(h))
            {
                if (net.GetNodeTemporalType(h) == 
                    Network.NodeTemporalType.Plate)
                {
                    int outcomeCount = net.GetOutcomeCount(h);
                    Console.WriteLine(
                        "Temporal beliefs for {0}:", net.GetNodeId(h));
                    double[] v = net.GetNodeValue(h);
                    for (int sliceIdx = 0; sliceIdx < sliceCount; sliceIdx++)
                    {
                        Console.Write("\tt={0}:", sliceIdx);
                        for (int i = 0; i < outcomeCount; i++)
                        {
                            Console.Write(
                                " {0}", v[sliceIdx * outcomeCount + i]);
                        }
                        Console.WriteLine();
                    }
                }
            }
            Console.WriteLine();
        }


        private static int CreateCptNode(
            Network net, String id, String name,
            String[] outcomes, int xPos, int yPos)
        {
            int handle = net.AddNode(Network.NodeType.Cpt, id);

            net.SetNodeName(handle, name);
            net.SetNodePosition(handle, xPos, yPos, 85, 55);

            int initialOutcomeCount = net.GetOutcomeCount(handle);
            for (int i = 0; i < initialOutcomeCount; i++)
            {
                net.SetOutcomeId(handle, i, outcomes[i]);
            }

            for (int i = initialOutcomeCount; i < outcomes.Length; i++)
            {
                net.AddOutcome(handle, outcomes[i]);
            }

            return handle;
        }
    }

}
